tracing-subscriber/src/fmt/fmt_subscriber.rs

```diff
@@ -1505,4 +1509,60 @@ mod test {
             actual.as_str()
         );
     }
+
+    // Because we need to modify an environment variable for these test cases,
+    // we do them all in a single test.
+    #[cfg(feature = "ansi")]
+    #[test]
+    fn subscriber_no_color() {
+        // save off the existing value of NO_COLOR
+        const NO_COLOR: &str = "NO_COLOR";
+        let saved_no_color = std::env::var(NO_COLOR);
+
+        let cases: Vec<(Option<&str>, bool)> = vec![
+            (Some("0"), false),   // any non-empty value disables ansi
+            (Some("off"), false), // any non-empty value disables ansi
+            (Some("1"), false),
+            (Some(""), true), // empty value does not disable ansi
+            (None, true),
+        ];
+
+        for (var, ansi) in cases {
+            if let Some(value) = var {
+                env::set_var(NO_COLOR, value);
+            } else {
+                env::remove_var(NO_COLOR);
+            }
+
+            let subscriber: Subscriber<()> = fmt::Subscriber::default();
+            assert_eq!(
+                subscriber.is_ansi, ansi,
+                "NO_COLOR={:?}; Subscriber::default().is_ansi should be {}",
+                var, ansi
+            );
+
+            // with_ansi should override any `NO_COLOR` value
+            let subscriber: Subscriber<()> = fmt::Subscriber::default().with_ansi(true);
+            assert!(
+                subscriber.is_ansi,
+                "NO_COLOR={:?}; Subscriber::default().with_ansi(true).is_ansi should be true",
+                var
+            );
+
+            // set_ansi should override any `NO_COLOR` value
+            let mut subscriber: Subscriber<()> = fmt::Subscriber::default();
+            subscriber.set_ansi(true);
+            assert!(
+                subscriber.is_ansi,
+                "NO_COLOR={:?}; subscriber.set_ansi(true); subscriber.is_ansi should be true",
+                var
+            );
+        }
+
+        if let Ok(value) = saved_no_color {
+            env::set_var(NO_COLOR, value);
+        } else {
+            env::remove_var(NO_COLOR);
+        }
```

Comment:
i think saving the original `NO_COLOR` env var and restoring it afterwards is a good idea. however, the way we're currently doing this, we only restore the previous value of the env var if the test _passes_. If any of the assertions panic, we won't hit the `if` block at the end, so the `NO_COLOR` env var won't be restored if we _really_ want to be good citizens about restoring the previous value, we might want to use a `Drop` guard, instead --- since local variables are _always_ dropped, even when unwinding, we can ensure that the previous value is restored even if the test panics. we might prefer something like this:
```suggestion

        const NO_COLOR: &str = "NO_COLOR";

        // Restores the previous value of the `NO_COLOR` env variable when
        // dropped.
        //
        // This is done in a `Drop` implementation, rather than just resetting
        // the value at the end of the test, so that the previous value is
        // restored even if the test panics.
        struct RestoreEnvVar(Result<String, env::VarError>);
        impl Drop for RestoreEnvVar {
            fn drop(&mut self) {
                match self.0 {
                    Ok(ref var) => env::set_var(NO_COLOR, var),
                    Err(_) => env::remove_var(NO_COLOR),
                }
            }
        }

        let _saved_no_color = RestoreEnvVar(env::var(NO_COLOR));

        let cases: Vec<(Option<&str>, bool)> = vec![
            (Some("0"), false),   // any non-empty value disables ansi
            (Some("off"), false), // any non-empty value disables ansi
            (Some("1"), false),
            (Some(""), true), // empty value does not disable ansi
            (None, true),
        ];

        for (var, ansi) in cases {
            if let Some(value) = var {
                env::set_var(NO_COLOR, value);
            } else {
                env::remove_var(NO_COLOR);
            }

            let subscriber: Subscriber<()> = fmt::Subscriber::default();
            assert_eq!(
                subscriber.is_ansi, ansi,
                "NO_COLOR={:?}; Subscriber::default().is_ansi should be {}",
                var, ansi
            );

            // with_ansi should override any `NO_COLOR` value
            let subscriber: Subscriber<()> = fmt::Subscriber::default().with_ansi(true);
            assert!(
                subscriber.is_ansi,
                "NO_COLOR={:?}; Subscriber::default().with_ansi(true).is_ansi should be true",
                var
            );

            // set_ansi should override any `NO_COLOR` value
            let mut subscriber: Subscriber<()> = fmt::Subscriber::default();
            subscriber.set_ansi(true);
            assert!(
                subscriber.is_ansi,
                "NO_COLOR={:?}; subscriber.set_ansi(true); subscriber.is_ansi should be true",
                var
            );
        }

    // dropping `_saved_no_color` will restore the previous value of 
    // `NO_COLOR`.
```

Reply:
Thanks for this Drop guard trick.  I knew there was a race around the environment variable, but I didn't know what was available to use as a workaround.